File: web/src/components/ingestion/harvest-options.ts
import type { FieldErrors, HarvestDefaults, HarvestOptionFields } from "@/types/domain";

const MIN_YEAR = 1990;
const MAX_YEAR = 2100;
const MAX_RECORDS = 50000;
const MAX_NAME_LENGTH = 120;
const MAX_CATEGORIES = 16;

/** Merges settings-derived defaults with the fields the user has actually touched. */
export function effectiveOptions(
  defaults: HarvestDefaults,
  overrides: Partial<HarvestOptionFields>,
): HarvestOptionFields {
  return {
    name: overrides.name,
    source: overrides.source ?? defaults.source,
    categories: overrides.categories,
    from: overrides.from ?? defaults.from,
    to: overrides.to ?? defaults.to,
    max: overrides.max ?? defaults.max,
  };
}

/** e.g. "arxiv · 2019–2026 · 2000" or "arxiv · 2019–2026 · 2000 · cs.LG, cs.AI". */
export function summarizeOptions(opts: HarvestOptionFields): string {
  const base = `${opts.source} · ${opts.from}–${opts.to} · ${opts.max}`;
  return opts.categories && opts.categories.length > 0 ? `${base} · ${opts.categories.join(", ")}` : base;
}

export interface ValidateContext {
  sources: readonly string[];
  hasSnapshotPath: boolean;
}

/** Mirrors internal/httpapi/harvest_options.go's resolveHarvest rules client-side. */
export function validateOptions(opts: HarvestOptionFields, ctx: ValidateContext): FieldErrors {
  const errors: FieldErrors = {};

  if (opts.name !== undefined && opts.name.length > MAX_NAME_LENGTH) {
    errors.name = `Name must be ${MAX_NAME_LENGTH} characters or fewer.`;
  }

  if (opts.source !== undefined) {
    if (ctx.sources.length > 0 && !ctx.sources.includes(opts.source)) {
      errors.source = `Unknown source "${opts.source}".`;
    } else if (opts.source === "kaggle" && !ctx.hasSnapshotPath) {
      errors.source = "Set the arXiv snapshot path in Settings first";
    }
  }

  if (opts.from !== undefined) {
    if (!Number.isInteger(opts.from) || opts.from < MIN_YEAR || opts.from > MAX_YEAR) {
      errors.from = `Year must be a whole number between ${MIN_YEAR} and ${MAX_YEAR}.`;
    }
  }
  if (opts.to !== undefined) {
    if (!Number.isInteger(opts.to) || opts.to < MIN_YEAR || opts.to > MAX_YEAR) {
      errors.to = `Year must be a whole number between ${MIN_YEAR} and ${MAX_YEAR}.`;
    }
  }
  if (!errors.from && !errors.to && opts.from !== undefined && opts.to !== undefined && opts.from > opts.to) {
    errors.to = "End year must be on or after the start year.";
  }

  if (opts.max !== undefined && (!Number.isInteger(opts.max) || opts.max < 1 || opts.max > MAX_RECORDS)) {
    errors.max = `Max records must be a whole number between 1 and ${MAX_RECORDS}.`;
  }

  if (opts.categories !== undefined) {
    if (opts.categories.length > MAX_CATEGORIES) {
      errors.categories = `At most ${MAX_CATEGORIES} categories.`;
    } else if (opts.categories.some((c) => /\s/.test(c))) {
      errors.categories = "A category code can't contain spaces.";
    }
  }
  if (!errors.categories && !errors.source && opts.source === "oai" && (opts.categories?.length ?? 0) === 0) {
    errors.categories = "oai requires at least one category.";
  }

  return errors;
}

/** Only the fields the user actually overrode — omitted fields use the server's defaults. */
export function toRequestFields(overrides: Partial<HarvestOptionFields>): HarvestOptionFields {
  const fields: HarvestOptionFields = {};
  if (overrides.name !== undefined && overrides.name.trim().length > 0) fields.name = overrides.name.trim();
  if (overrides.source !== undefined) fields.source = overrides.source;
  if (overrides.categories !== undefined && overrides.categories.length > 0) {
    fields.categories = overrides.categories;
  }
  if (overrides.from !== undefined) fields.from = overrides.from;
  if (overrides.to !== undefined) fields.to = overrides.to;
  if (overrides.max !== undefined) fields.max = overrides.max;
  return fields;
}

File: internal/httpapi/harvest_options.go
package httpapi

import (
	"net/url"
	"slices"
	"strconv"
	"strings"

	"github.com/yothgewalt/reharvester/internal/harvest"
)

// harvestMaxCap bounds a requested max the same way settings.ValidMax bounds
// the configured one; a request over it is clamped rather than rejected.
const harvestMaxCap = 50000

// harvestNameLimit and harvestCategoriesLimit bound the two fields with no
// natural numeric range.
const (
	harvestNameLimit       = 120
	harvestCategoriesLimit = 16
)

// harvestOptions are the fields POST /api/v1/harvest/init accepts alongside
// keywords/abstract/pdf. A zero value in From, To or Max, or a blank Source,
// means "use the default".
type harvestOptions struct {
	Name       string
	Source     string
	Categories []string
	From       int
	To         int
	Max        int
}

// harvestDefaults are what resolveHarvest falls back to for a blank option.
type harvestDefaults struct {
	Source          string
	Max             int
	HasSnapshotPath bool
}

// resolveHarvest fills in opts against def and validates the result. year
// anchors From/To when both are blank (year-7..year). errs is keyed by the
// same field names the JSON and multipart request bodies use, so a handler
// can return it directly as the response's "errors" object.
func resolveHarvest(opts harvestOptions, def harvestDefaults, year int) (q harvest.Query, source, name string, errs map[string]string) {
	errs = map[string]string{}

	source = strings.ToLower(strings.TrimSpace(opts.Source))
	if source == "" {
		source = def.Source
	} else if !slices.Contains(harvest.Sources, source) {
		errs["source"] = "must be one of " + strings.Join(harvest.Sources, ", ")
	}
	if source == harvest.SourceKaggle && !def.HasSnapshotPath {
		errs["source"] = "the kaggle source needs an arXiv snapshot path — set one in Settings"
	}

	name = strings.TrimSpace(opts.Name)
	if len(name) > harvestNameLimit {
		errs["name"] = "must be 120 characters or fewer"
	}

	cats := opts.Categories
	switch {
	case len(cats) > harvestCategoriesLimit:
		errs["categories"] = "at most 16 categories"
	case source == harvest.SourceOAI && len(cats) == 0:
		errs["categories"] = "the oai source needs at least one category"
	}

	from, to := opts.From, opts.To
	if from == 0 {
		from = year - 7
	}
	if to == 0 {
		to = year
	}
	if from > to {
		errs["from"] = "from year must not be after to year"
	}

	max := opts.Max
	switch {
	case max < 0:
		errs["max"] = "must not be negative"
	case max == 0:
		max = def.Max
	case max > harvestMaxCap:
		max = harvestMaxCap
	}

	if len(errs) > 0 {
		return harvest.Query{}, "", "", errs
	}
	return harvest.Query{Categories: cats, From: from, To: to, Max: max}, source, name, nil
}

// optionsFromForm reads harvestOptions out of a multipart form's values.
// errs carries a message for any field that could not be parsed as a whole
// number; a blank field never errors, since blank means "use the default".
func optionsFromForm(v url.Values) (harvestOptions, map[string]string) {
	errs := map[string]string{}
	opts := harvestOptions{
		Name:       strings.TrimSpace(v.Get("name")),
		Source:     strings.TrimSpace(v.Get("source")),
		Categories: splitComma(v.Get("categories")),
		From:       parseFormInt(v.Get("from"), "from", errs),
		To:         parseFormInt(v.Get("to"), "to", errs),
		Max:        parseFormInt(v.Get("max"), "max", errs),
	}
	return opts, errs
}

func parseFormInt(s, field string, errs map[string]string) int {
	s = strings.TrimSpace(s)
	if s == "" {
		return 0
	}
	n, err := strconv.Atoi(s)
	if err != nil {
		errs[field] = "must be a whole number"
		return 0
	}
	return n
}
